src/expr/src/expr/expr_array_concat.rs

```diff
@@ -0,0 +1,488 @@
+// Copyright 2022 Singularity Data
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::sync::Arc;
+
+use itertools::Itertools;
+use risingwave_common::array::{ArrayRef, DataChunk, ListValue, Row};
+use risingwave_common::types::{to_datum_ref, DataType, Datum, DatumRef, ScalarRefImpl};
+use risingwave_pb::expr::expr_node::{RexNode, Type};
+use risingwave_pb::expr::ExprNode;
+
+use crate::expr::{build_from_prost as expr_build_from_prost, BoxedExpression, Expression};
+use crate::{bail, ensure, ExprError, Result};
+
+#[derive(Debug)]
+pub enum Operation {
+    ConcatArray,
+    AppendArray,
+    PrependArray,
+    AppendValue,
+    PrependValue,
+}
+
+#[derive(Debug)]
+pub struct ArrayConcatExpression {
+    return_type: DataType,
+    left: BoxedExpression,
+    right: BoxedExpression,
+    op: Operation,
+}
+
+impl ArrayConcatExpression {
+    pub fn new(
+        return_type: DataType,
+        left: BoxedExpression,
+        right: BoxedExpression,
+        op: Operation,
+    ) -> Self {
+        Self {
+            return_type,
+            left,
+            right,
+            op,
+        }
+    }
+
+    /// Concatenates two arrays with same data type.
+    /// The behavior is the same as PG.
+    ///
+    /// Examples:
+    /// - `select array_cat(array[66], array[123]);` => `[66,123]`
+    /// - `select array_cat(array[66], null::int[]);` => `[66]`
+    /// - `select array_cat(null::int[], array[123]);` => `[123]`
+    fn concat_array(left: DatumRef, right: DatumRef) -> Datum {
+        match (left, right) {
+            (None, right) => right.map(ScalarRefImpl::into_scalar_impl),
+            (left, None) => left.map(ScalarRefImpl::into_scalar_impl),
+            (Some(ScalarRefImpl::List(left)), Some(ScalarRefImpl::List(right))) => Some(
+                ListValue::new(
+                    left.values_ref()
+                        .into_iter()
+                        .chain(right.values_ref().into_iter())
+                        .map(|x| x.map(ScalarRefImpl::into_scalar_impl))
+                        .collect(),
```

Comment:
~~Can the `iter` be optimized out?~~ Seems not possible now, because `ListValue` is a naive `Vec`

Reply:
Seems not possible because `ListValue` is simply `Vec<Datum>`